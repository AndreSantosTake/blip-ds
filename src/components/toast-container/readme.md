# toast-container



<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [bds-toast](../toast)

### Graph
```mermaid
graph TD;
  bds-toast --> bds-toast-container
  style bds-toast-container fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
