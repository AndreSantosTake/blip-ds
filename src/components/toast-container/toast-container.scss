:host {
  position: fixed;
  bottom: 40px;
  left: 32px;
  display: flex;
  flex-direction: column-reverse;
}
