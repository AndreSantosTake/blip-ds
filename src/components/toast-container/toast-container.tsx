import { Component, h, Host } from '@stencil/core';

@Component({
  tag: 'bds-toast-container',
  styleUrl: 'toast-container.scss',
  shadow: true,
})
export class BdsToastContainer {
  render() {
    return (
      <Host>
        <slot></slot>
      </Host>
    );
  }
}
