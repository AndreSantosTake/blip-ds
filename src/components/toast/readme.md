# bds-toast



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                                                                                          | Type                                            | Default     |
| ------------ | ------------- | ---------------------------------------------------------------------------------------------------- | ----------------------------------------------- | ----------- |
| `actionType` | `action-type` | ActionType. Defines if the button should have a button or an icon. Can be one of:  'icon', 'button'; | `"button" \| "icon"`                            | `'icon'`    |
| `buttonText` | `button-text` | If the action type is button, this will be the text of the button:                                   | `string`                                        | `undefined` |
| `icon`       | `icon`        | used for add the icon. Uses the bds-icon component.                                                  | `string`                                        | `null`      |
| `text`       | `text`        | The text content of the component:                                                                   | `string`                                        | `undefined` |
| `toastTitle` | `toast-title` | The title of the component:                                                                          | `string`                                        | `undefined` |
| `variant`    | `variant`     | Variant. Defines the color of the toast. Can be one of:  'system', 'error', 'success', 'warning';    | `"error" \| "success" \| "system" \| "warning"` | `'system'`  |


## Methods

### `open() => Promise<void>`

Can be used outside to open the toast

#### Returns

Type: `Promise<void>`




## Dependencies

### Depends on

- [bds-icon](../icon)
- [bds-typo](../typo)
- [bds-button](../button)

### Graph
```mermaid
graph TD;
  bds-toast --> bds-icon
  bds-toast --> bds-typo
  bds-toast --> bds-button
  bds-button --> bds-icon
  bds-button --> bds-typo
  style bds-toast fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
