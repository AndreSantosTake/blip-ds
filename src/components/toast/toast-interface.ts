export type ActionType = 'button' | 'icon';

export type VariantType = 'system' | 'error' | 'success' | 'warning';

export type ButtonActionType = 'close' | 'custom';

export type IconVariantMap = { [key in VariantType]: string };

export type CreateToastType = {
  buttonAction?: ButtonActionType;
  buttonText?: string;
  toastText: string;
  toastTitle: string;
  icon?: string;
  actionType?: ActionType;
  variant?: VariantType;
  duration?: number;
};
