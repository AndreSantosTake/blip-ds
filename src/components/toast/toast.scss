@import "mixins";

@mixin inherit-color() {
  color: inherit;
  background-color: inherit;
}
:host {
  .show {
    opacity: 1;
    animation: toastAnimationFadeIn 1s;
  }
  .hide {
    transition: all 1s;
    animation: toastAnimationFadeOut 0.5s;
  }
}

.toast {
  ::before {
    display: block;
  }

  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  border-radius: 8px;
  box-shadow: $shadow-toast;
  color: $color-snow-white;
  opacity: 0;
  margin-top: 16px;

  &--action--icon {
    min-width: 440px;
    max-width: 440px;
    padding: 8px 20px;
    
    bds-icon-button {
      height: 32px;
    }

    @media (max-width: 780px) {
      min-width: 220px;
      width: 95%;
      margin: 16px auto 0px auto;
    }
  }

  &--action--button {
    min-width: 440px;
    max-width: 456px;
    padding: 10px 20px;

    @media (max-width: 780px) {
      min-width: 220px;
      width: 95%;
      margin: 16px auto 0px auto;
    }
  }

  &--system {
    background: $color-primary-main;
  }

  &--error {
    background: $color-secondary-delete;
  }

  &--success {
    background: $color-secondary-true;
  }

  &--warning {
    background: $color-secondary-cheetos;
  }

  bds-icon {
    margin-right: 8px;
    display: flex;
    align-items: center;
  }

  &__content {
    height: 100%;
    width: 100%;
    align-items: flex-start;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-right: 8px;
  }

  &__action {
    display: flex;
    align-items: center;

    button {
      background: transparent;
      border: none;
      outline: none;
      padding: 0;
    }

    &__button {
      white-space: nowrap;
    }

    &__icon {
      display: flex !important;
      align-items: center;

      bds-icon {
        color: $color-snow-white;
        border-radius: 4px;
        cursor: pointer;
      }
    }
  }
}

@keyframes toastAnimationFadeIn {
  0% {
    opacity: 0;
    left: -200px;
  }
  50% {
    opacity: 0.9;
    left: 1px;
  }
  100% {
    opacity: 1;
  }
}

@keyframes toastAnimationFadeOut {
  0% {
    opacity: 1;
  }
  30% {
    max-height: 60px;
  }
  80% {
    opacity: 0;
    max-height: 30px;
  }
  100% {
    max-height: 0px;
  }
}
