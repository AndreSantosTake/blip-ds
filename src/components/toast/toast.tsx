import {
  Component,
  ComponentInterface,
  h,
  Prop,
  Method,
  Element,
  Event,
  EventEmitter,
} from "@stencil/core";

export type ActionType = "button" | "icon";

export type VariantType = "system" | "error" | "success" | "warning";

export type ButtonActionType = "close" | "custom";

export type CreateToastType = {
  toastContainer: HTMLToastContainerElement;
  toastElement: HTMLBdsToastElement;
  actionType: ActionType;
  buttonAction: ButtonActionType;
  buttonText?: string;
  icon?: string;
  toastText?: string;
  toastTitle?: string;
  variant?: VariantType;
  duration: number;
}

@Component({
  tag: "bds-toast",
  styleUrl: "toast.scss",
  shadow: true,
})
export class BdsToast implements ComponentInterface {
  @Element() el: HTMLElement;
  /**
   * used for add the icon. Uses the bds-icon component.
   */
  @Prop({ reflect: true }) icon?: string = null;
  /**
   * ActionType. Defines if the button should have a button or an icon. Can be one of:
   * 'icon', 'button';
   */
  @Prop() actionType: ActionType = "button";
  /**
   * Variant. Defines the color of the toast. Can be one of:
   * 'system', 'error', 'success', 'warning';
   */
  @Prop() variant: VariantType = "system";
  /**
   * The title of the component:
   */
  @Prop() toastTitle: string;
  /**
   * The text content of the component:
   */
  @Prop() toastText: string;
  /**
   * If the action type is button, this will be the text of the button:
   */
  @Prop() buttonText: string;
  /**
   * Time to close the toast in seconds
   * 0 = no auto close (default value)
   */
  @Prop() duration = 0;
  /**
   * Define an action to the button toast. Can be one of:
   * 'close', 'custom';
   * if the action type is set to close, the button will close automatically.
   * if the action type is set to custom, a function need to be passed when the toastButtonClick is emitted.
   */
  @Prop() buttonAction: ButtonActionType = "close";
  /**
   * Controls the open event of the component:
   */
  @Prop() show = false;
    /**
   * Controls the hide event of the component:
   */
  @Prop() hide = false;
  /**
   * Event used to execute some action when the action button on the toast is clicked
   */
  @Event() toastButtonClick!: EventEmitter;
  /**
   * Sends an event to be used when creating an action when clicking the toast button
   */
  private _buttonClickHandler = (event) => {
    const element = event.path.find((item) => item.localName === "bds-toast");

    if (this.buttonAction === "close") this.close(event);
    else this.toastButtonClick.emit({ element });
  };

  /**
   * Can be used outside to open the toast
   */
  @Method()
  async create({
    toastContainer,
    toastElement,
    actionType,
    buttonAction,
    buttonText,
    icon,
    toastText,
    toastTitle,
    variant,
    duration,
  }: CreateToastType) {
    toastElement.actionType = actionType || 'button';
    toastElement.buttonAction = buttonAction;
    toastElement.buttonText = buttonText;
    toastElement.icon = icon || 'info';
    toastElement.toastText= toastText;
    toastElement.toastTitle = toastTitle;
    toastElement.variant = variant || 'system';
    toastElement.duration = duration * 1000 || 0;

    toastContainer.appendChild(toastElement);
    toastElement.show = true;

    if (toastElement.duration > 0) {
      setTimeout(() => {
        toastElement.hide = true;
        setTimeout(() => {
          toastElement.remove();
        }, 400);
      }, toastElement.duration);
    }
  }

  /**
   * Can be used outside the component to close the toast
   */
  @Method()
  async close(event) {
    const element = event.path.find((item) => item.localName === "bds-toast");
    const tElement = await element.shadowRoot.childNodes[1];

    tElement.classList.remove('show');
    tElement.classList.add('hide');

    setTimeout(() => {
      element.remove();
    }, 400);
  }

  render() {
    return (
      <div
        class={{
          toast: true,
          [`toast--${this.variant}`]: true,
          [`toast--action--${this.actionType}`]: true,
          show: this.show,
          hide: this.hide
        }}
      >
        {this.icon && (
          <bds-icon
            theme="outline"
            size="x-large"
            color="#fff"
            name={this.icon}
          />
        )}
        <div class="toast__content">
          {this.toastTitle && (
            <bds-typo variant="fs-14" bold="bold">
              {this.toastTitle}
            </bds-typo>
          )}
          {this.toastText && <bds-typo variant="fs-14">{this.toastText}</bds-typo>}
        </div>
        <div
          class={{
            "toast__action": true,
            [`toast__action__${this.actionType}`]: true,
          }}
        >
          {this.actionType === "button" ? (
            <bds-button
              variant="secondary--white"
              onClick={(event) => this._buttonClickHandler(event)}
              size="short"
            >
              {this.buttonText}
            </bds-button>
          ) : (
            <bds-icon-button
              size="short"
              variant="secondary--white"
              icon="close"
              onClick={(event) => this.close(event)}
            />
          )}
        </div>
      </div>
    );
  }
}
