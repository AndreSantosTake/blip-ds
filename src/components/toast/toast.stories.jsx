import React from "react";
import { withKnobs, text, select } from "@storybook/addon-knobs";

import readme from "./readme.md";

export default {
  title: "Toast",
  decorators: [withKnobs],
  parameters: {
    notes: { markdown: readme },
  },
};

export const toastSystem = () => {
  const variantOptions = {
    System: "system",
    Error: "error",
    Success: "success",
    Warning: "warning",
  };
  const actionTypeOptions = { Button: "button", Icon: "icon" };
  const buttonActionOptions = { Close: "close", Custom: "custom" };

  const showToast = async () => {
    const newtoast = document.createElement("bds-toast");
    const toastContainer = document.querySelector("bds-toast-container")
      .shadowRoot;
    toastContainer.appendChild(newtoast);

    const icon = text("icon", undefined);
    const variant = select("variant", variantOptions);
    const actionType = select("action-type", actionTypeOptions);
    const toastTitle = text("toast-title", "Toast Title");
    const toastText = text("toast-text", "Lorem ipsum");
    const buttonText = text("button-text", "Cancelar");
    const duration = text("duration", 0);
    const buttonAction = select("button-action", buttonActionOptions);

    await newtoast.create({
      variant,
      actionType,
      toastTitle,
      buttonText,
      buttonAction,
      toastText,
      variant,
      duration,
      icon,
    });
  };

  return (
    <>
      <bds-toast-container></bds-toast-container>
      <bds-button onClick={() => showToast()}>Show toast</bds-button>
    </>
  );
};
