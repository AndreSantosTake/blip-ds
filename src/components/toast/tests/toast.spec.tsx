import { newSpecPage } from "@stencil/core/testing";
import { BdsToast } from "../toast";
import { ToastContainer } from "../../toast-container/toast-container";


describe("bds-toast", () => {
  const getToast = async () => {
    return await newSpecPage({
      components: [BdsToast, ToastContainer],
      html: `
      <toast-container></toast-container>
      <bds-toast title="teste" toast-title="titulo" toast-text="texto" icon="close" action-type="button" buttona-action="close"></bds-toast>`,
    })
  };

  it("should render the component", async () => {
    const staticToastPage = await getToast();
    expect(staticToastPage.root).toMatchSnapshot();
  });
});
