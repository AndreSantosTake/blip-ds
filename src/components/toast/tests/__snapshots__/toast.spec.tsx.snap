// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bds-toast should render the component 1`] = `
<toast-container>
  <mock:shadow-root>
    <div>
      <slot></slot>
    </div>
  </mock:shadow-root>
</toast-container>
`;
